### PR TITLE
When deleting a collection delete its related data thoroughly

### LINKED ISSUE #3848: Collection tree file not deleted when collection deleted
## Bug Description

When you delete a collection via the CP, the "tree" `.yaml` file is not deleted.

## How to Reproduce

1. Add Collection
2. Add entries
3. Give it some structure (ordered)
4. Delete collection

## Environment

Statamic 3.1.23 Pro
Laravel 8.46.0
PHP 8.0.7
aryehraber/statamic-captcha 1.5.1
aryehraber/statamic-uuid 2.0.3
edalzell/forma 0.8.12
jackabox/statamic-shopify 1.4.3
statamic/seo-pro 2.2.1
transformstudios/events 2.1.4
transformstudios/front 1.2.1
transformstudios/magic-link dev-main
transformstudios/review 3.0.4
transformstudios/simple dev-main
transformstudios/stripe dev-main
transformstudios/takeover-redirects dev-main

**Install method (choose one):**
- Starter kit

### PR DESCRIPTION
**Pull request by arthurperton**
Fixes #3848 and #3890.

* Deletes entry localizations, if any
* Deletes the trees, if any

### CODE REVIEW

File: src/Entries/Collection.php [hunk 1]
```diff
@@ -602,6 +603,12 @@ public function delete()
     {
         $this->queryEntries()->get()->each->delete();
 
+        if ($this->structure()) {
+            $this->structure()->trees()->each(function ($tree) {
+                app(CollectionTreeRepository::class)->delete($tree);
```
Comment (jelleroorda) [inline src/Entries/Collection.php:608]: Can't you just call `$tree->delete()` there? It'd also dispatch the deleted event.

Comment (arthurperton) [reply src/Entries/Collection.php:608]: Thanks Jelle, sounds like an improvement. I’ll look into it soon.

Comment (jasonvarga) [reply src/Entries/Collection.php:608]: Yeah you should be able to to `$this->structure()->trees()->each->delete();`

**Review by jasonvarga — CHANGES_REQUESTED** [review]
I'd love if #3891 could be combined with this PR. As I mentioned there, it's such similar stuff happening right next to the other, I feel like it'd be better tested together.

**Comment by arthurperton** [comment]
Yeah totally, I found the one issue while solving the other. The fixes are combined now into this PR.

**Comment by arthurperton** [comment]
@jasonvarga I wonder what you mean in this [comment](https://github.com/statamic/cms/pull/3902#issuecomment-872357424). What is it that needs doing?

**Comment by jasonvarga** [comment]
I think I just meant once it was merged.

**Review by jasonvarga — CHANGES_REQUESTED** [review]
I get an error when I delete the collection when using multisite.

```
[2021-07-14 22:48:04] local.ERROR: Call to a member function pages() on null {"userId":"3ad6235d-8556-42dd-a500-9d7cc1373bd0","exception":"[object] (Error(code: 0): Call to a member function pages() on null at /users/jason/code/statamic/three/cms/src/Entries/Entry.php:167)
[stacktrace]
#0 /Users/jason/Code/statamic/three/sandbox/vendor/laravel/framework/src/Illuminate/Support/helpers.php(263): Statamic\\Entries\\Entry->Statamic\\Entries\\{closure}(Object(Statamic\\Structures\\CollectionTree))
#1 /users/jason/code/statamic/three/cms/src/Entries/Entry.php(171): tap(Object(Statamic\\Structures\\CollectionTree), Object(Closure))
#2 /Users/jason/Code/statamic/three/sandbox/vendor/laravel/framework/src/Illuminate/Support/helpers.php(263): Statamic\\Entries\\Entry->Statamic\\Entries\\{closure}(Object(Statamic\\Structures\\CollectionStructure))
#3 /users/jason/code/statamic/three/cms/src/Entries/Entry.php(172): tap(Object(Statamic\\Structures\\CollectionStructure), Object(Closure))
#4 /users/jason/code/statamic/three/cms/src/Entries/Collection.php(605): Statamic\\Entries\\Entry->delete()
#5 /Users/jason/Code/statamic/three/sandbox/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php(235): Statamic\\Entries\\Collection->Statamic\\Entries\\{closure}(Object(Statamic\\Entries\\Entry), 2)
#6 /users/jason/code/statamic/three/cms/src/Entries/Collection.php(606): Illuminate\\Support\\Collection->each(Object(Closure))
```

**Comment by arthurperton** [comment]
Thanks I'll check that.

**Comment by arthurperton** [comment]
> I get an error when I delete the collection when using multisite.

The last commit fixes the error. But when using multisite, the tree files are not deleted (well actually they are, but the get created again) and remain with `{ }` inside. So this still needs some work.

**Comment by jasonvarga** [comment]
Closing as this has been sitting as a draft for a while. Feel free to reopen when it's ready or there's a better solution. Thanks!

**Pull request closed without merging**